_posts/2018-07-23-performance-profiling-with-craft-cms.md

```diff
@@ -0,0 +1,117 @@
+
+A relation in craft has a source and a target. The source has a relation field and selects the other element. The target is selected by the source.
+[Relations | Craft 2 Documentation](https://docs.craftcms.com/v2/relations.html#terminology)
+> It’s important to grasp those terms, as they are relevant to the templating side of things
+
+
+In Craft, [elements](https://docs.craftcms.com/v2/plugins/working-with-elements.html) are:  
+* Assets
+* Categories
+* Entries
+* Global Sets
+* Matrix Blocks
+* Tags
+* Users 
+<br><br>
+
+It takes some steps to implement Eager Loading.
+This example shows you how you Eager Load an imageField:
+
+#### No Eager Loading
+{% highlight twig %}
+  {% raw %}
+    {% set entries = craft.entries().section('info').all() %}
+
+    {% for entry in entries %}
+      {% set image = entry.imageField.one()%}
+      {% if image %}
+        <img src="{{ image.url }}" alt="{{ image.title }}">
+      {% endif %}
+    {% endfor %}
+  {% endraw %}
+{% endhighlight %}
+
+#### With Eager Loading
+{% highlight twig %}
+  {% raw %}
+    {% set 
+      entries = craft.entries()
+      .section('info') // 1. fetch entries from info section
+      .with(['imageField']) // 2. what should be Eager Loaded?
+      .all() 
+    %} 
+
+    {% for entry in entries %}
+      {% set image = entry.imageField[0] %} // 3. fetch the image
+      {% if image %}
+        <img src="{{ image.url }}" alt="{{ image.title }}">
+      {% endif %}
+    {% endfor %}
+  {% endraw %}
+{% endhighlight %}
+
+There are 2 major things you will notice:
+* While selecting all entries from a section (`info`), you need to tell Craft, which fields you are going to use. In this case, we will need the `imageField`. Please note, that this can be a comma separated list (Array syntax) and you can have more than one field here.
+* We also need to change the image selector from `.one()` to `[0]`
+
+With Eager Loading, no matter how many entries you have to fetch, Craft only needs 3 database queries instead of N+1 (`N` = number of entries + `1` = initial entries query).
+
+* Make sure you also read through the [Craft 3 Documentation](https://docs.craftcms.com/v3/dev/eager-loading-elements.html#eager-loading-image-transform-indexes)
+* If you want to dive deeper, here is a really good & detailed read about [Eager loading](https://nystudio107.com/blog/speed-up-your-craft-cms-templates-with-eager-loading).
+
+### Conclusion
+Well of course this heavily depends on the page you are serving. But in general it is a good idea to have  the first contentful paint (FCP) below 2s. If you are higher than that, does not mean you are serving a bad website, but this means that your user will lose interest very quickly. As you can see in the illustration, the bounce rate increases slowly after 2 seconds and rapidly above 3 seconds.
```

Comment:
Of course the overall performance heavily depends on the page you are serving, but in general...